jobqueue_features/tests/test_pbs_ci.py

```diff
@@ -0,0 +1,320 @@
+from __future__ import print_function
+
+from unittest import TestCase
+import os
+
+import pytest
+
+from jobqueue_features.clusters_controller import (
+    clusters_controller_singleton as controller,
+)
+
+from jobqueue_features import (
+    mpi_wrap,
+    MPIEXEC,
+    OPENMPI,
+    on_cluster,
+    mpi_task,
+    which,
+    CustomPBSCluster,
+    get_task_mpi_comm
+)
+
+# Use logging if there are hard to see issues in the CI
+
+# import logging
+# logging.basicConfig(format="%(levelname)s:%(message)s", level=logging.DEBUG)
+
+
+class TestCIPBS(TestCase):
+    def setUp(self):
+        # Kill any existing clusters
+        controller._close()
+
+        self.number_of_processes_per_node = 2
+        # Really hard to get srun in CI, so use mpiexec to keep things simple
+        self.launcher = MPIEXEC
+
+        self.run_tests = False
+        if which(OPENMPI["launcher"]) is not None:
+            print(
+                "Found {} so assuming we have PBS, running MPI test (with {})".format(
+                    OPENMPI["launcher"], self.launcher
+                )
+            )
+            self.run_tests = True
+
+        self.common_kwargs = {
+            "interface": None,
+            "walltime": "00:01:00",
+            "cores_per_node": 2,
+            "minimum_cores": 2,
+            "hyperthreading_factor": 1,
+            "ntasks_per_node": 2,
+            "memory": "2 GB",
+            "mpi_mode": True,
+            "env_extra": [
+                "export OMPI_ALLOW_RUN_AS_ROOT_CONFIRM=1",
+                "export OMPI_ALLOW_RUN_AS_ROOT=1",
+            ],
+            "mpi_launcher": self.launcher,
+            "local_directory": "/tmp",
+            "queue": "workq",
+        }
+
+        self.executable = "python"
+        self.script_path = os.path.abspath(
+            os.path.join(os.path.dirname(__file__), "resources", "helloworld.py")
+        )
+
+    @pytest.mark.env("pbs")
+    def test_single_mpi_wrap(self):
+        #
+        # Assume here we have srun support
+        if self.run_tests:
+            controller._close()
+
+            # Create the cluster
+            nodes = 2
+            fork_pbs_cluster = CustomPBSCluster(
+                name="fork_cluster", fork_mpi=True, nodes=nodes, **self.common_kwargs
+            )
+
+            # Create the function that wraps tasks for this cluster
+            @on_cluster(cluster=fork_pbs_cluster, jobs=1)
```

Comment:
These tests are a pretty good example of how to use this library and they are basically identical for PBS and SLURM, can we refactor them to avoid a lot of the duplication?